src/main/java/org/wildfly/security/auth/provider/oauth2/OAuth2Util.java

```diff
@@ -0,0 +1,190 @@
+import java.util.Map;
+
+/**
+ * An utility class providing static methods to connect to standard OAuth2 HTTP endpoints.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+class OAuth2Util {
+
+    /**
+     * Introspects an OAuth2 Access Token using a RFC-7662 compatible endpoint.
+     *
+     * @param tokenIntrospectionUrl an {@link URL} pointing to a RFC-7662 compatible endpoint
+     * @param clientId the identifier of a client within the OAUth2 Authorization Server
+     * @param clientSecret the secret of the client
+     * @param token the access token to introspect
+     * @param sslContext the ssl context
+     * @param hostnameVerifier the hostname verifier
+     * @return a @{JsonObject} representing the response from the introspection endpoint or null if
+     */
+    static JsonObject introspectAccessToken(URL tokenIntrospectionUrl, String clientId, String clientSecret, String token, SSLContext sslContext, HostnameVerifier hostnameVerifier) throws RealmUnavailableException {
+        Assert.checkNotNullParam("clientId", clientId);
+        Assert.checkNotNullParam("clientSecret", clientSecret);
+        Assert.checkNotNullParam("token", token);
+
+        HttpURLConnection connection = null;
+
+        try {
+            connection = openConnection(tokenIntrospectionUrl, sslContext, hostnameVerifier);
+
+            HashMap<String, String> parameters = new HashMap<>();
+
+            parameters.put("token", token);
+            parameters.put("token_type_hint", "access_token");
+
+            byte[] params = buildParameters(parameters);
+
+            connection.setDoOutput(true);
+            connection.setRequestMethod("POST");
+            connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+            connection.setRequestProperty("Content-Length", String.valueOf(params.length));
+            connection.setRequestProperty("Authorization", "Basic " + CodePointIterator.ofString(clientId + ":" + clientSecret).asUtf8().base64Encode().drainToString());
+
+            try (OutputStream outputStream = connection.getOutputStream()) {
+                outputStream.write(params);
+            }
+
+            try (InputStream inputStream = new BufferedInputStream(connection.getInputStream())) {
+                return Json.createReader(inputStream).readObject();
+            }
+        } catch (IOException ioe) {
+            if (connection != null && connection.getErrorStream() != null) {
+                InputStream errorStream = connection.getErrorStream();
+
+                try (BufferedReader reader = new BufferedReader(new InputStreamReader(errorStream))) {
+                    StringBuffer response = reader.lines().reduce(new StringBuffer(), StringBuffer::append, (buffer1, buffer2) -> buffer1);
+                    ElytronMessages.log.errorf(ioe, "Unexpected response from token introspection endpoint [%s]. Response: [%s]", tokenIntrospectionUrl, response);
```

Comment:
I don't think we need to trigger an exception here. We may have non OK responses from server that we just need to fail the evidence verification and log the error in logs.